src/wallet_sync.rs

```diff
@@ -91,127 +93,212 @@ const WATCH_ONLY_SWAPCOIN_LABEL: &str = "watchonly_swapcoin_label";
+                        index,
+                        redeemscript,
+                        self.funding_amount,
+                        SigHashType::All,
+                    )[..],
+                )
+                .unwrap();
+
+                let sig_mine = secp.sign(&sighash, &self.my_privkey);
+                let sig_other = secp.sign(&sighash, &self.other_privkey.unwrap());
+
+                let (sig_first, sig_second) =
+                    if my_pubkey.serialize()[..] < self.other_pubkey.serialize()[..] {
+                        (sig_mine, sig_other)
+                    } else {
+                        (sig_other, sig_mine)
+                    };
+
+                input.witness.push(Vec::new()); //first is multisig dummy
+                input.witness.push(sig_first.serialize_der().to_vec());
+                input.witness.push(sig_second.serialize_der().to_vec());
+                input.witness[1].push(SigHashType::All as u8);
+                input.witness[2].push(SigHashType::All as u8);
+                input.witness.push(redeemscript.to_bytes());
+                Ok(())
+            }
+        }
+    };
 }
 
+impl_wallet_swapcoin!(IncomingSwapCoin);
+impl_wallet_swapcoin!(OutgoingSwapCoin);
+
 impl Wallet {
     pub fn print_wallet_key_data(&self) {
         println!(
             "master key = {}, external_index = {}",
             self.master_key, self.external_index
         );
-        for (multisig_redeemscript, swapcoin) in &self.swap_coins {
-            println!(
-                "{} {}:{} {}",
-                Address::p2wsh(multisig_redeemscript, NETWORK),
-                swapcoin.contract_tx.input[0].previous_output.txid,
-                swapcoin.contract_tx.input[0].previous_output.vout,
-                if swapcoin.other_privkey.is_some() {
-                    "  known"
-                } else {
-                    "unknown"
-                }
-            )
+
+        for (multisig_redeemscript, swapcoin) in &self.incoming_swap_coins {
+            Self::print_script_and_coin(multisig_redeemscript, swapcoin);
+        }
+        for (multisig_redeemscript, swapcoin) in &self.outgoing_swap_coins {
+            Self::print_script_and_coin(multisig_redeemscript, swapcoin);
         }
-        println!("swapcoin count = {}", self.swap_coins.len());
+        println!(
+            "swapcoin count = {}",
+            self.incoming_swap_coins.len() + self.outgoing_swap_coins.len()
+        );
+    }
+
+    fn print_script_and_coin(script: &Script, coin: &dyn SwapCoin) {
+        let contract_tx = coin.get_contract_tx();
+        println!(
+            "{} {}:{} {}",
+            Address::p2wsh(script, NETWORK),
+            contract_tx.input[0].previous_output.txid,
+            contract_tx.input[0].previous_output.vout,
+            if coin.get_other_privkey().is_some() {
+                "  known"
+            } else {
+                "unknown"
+            }
+        )
```

Comment:
We can avoid using SwapCoin.get_other_privkey() here if we retain knowledge of the type (either IncomingSwapCoin or OutgoingSwapCoin). OutgoingSwapCoin will never have the field `other_privkey`. Probably this function `print_script_and_coin` should only take IncomingSwapCoins.

Reply:
> Probably this function print_script_and_coin should only take IncomingSwapCoins.

I think it's important to print all the coins in this context, and added an `is_known` function to `SwapCoin`. For `IncomingSwapCoin` is tests the presence of `hash_preimage` or `other_privkey`, and for `OutgoingSwapCoin` only tests for `hash_preimage`.